### PR TITLE
Task class

### PR DESCRIPTION
**Pull request by matthewhanson**
This draft Task class (see #42) is derived from the Cirrus task template: https://github.com/cirrus-geo/cirrus-task-template

### CODE REVIEW

**Comment by matthewhanson** [comment]
I was creating a new task for Cirrus, so I decided to just expand on the Task class here and get a working version.

I've also added a task example in the tests directory that show a basic copy task, now far simpler than having all the previous boilerplate.

**Comment by jkeifer** [comment]
As you know I really like this idea. I wish I had more time to review and test out some of my thoughts, but here's a couple high-level comments:

* I think the base class here is not `Task`, but `Lambda`, and `Task` should be a subclass of `Lambda`. Then we'd also have `Feeder` and `Function` subclasses. I say that because things like logging, event parsing, and args/cli functions would be common across all of these "lambda" types.
* I would look into using the `__init_subclass__` method on that base `Lambda` class to register the `handler` method as the function `lambda_handler` within the subclass's module. I might have this totally wrong, but it could be something like this:

```
    def __init_subclass__(cls, abstract=False, **kwargs):
        super().__init_subclass__(**kwargs)
        if abstract:
            return
        module = cls.__module__
        module.lambda_handler = cls.handler
```

**Comment by matthewhanson** [comment]
Good point on having another base class, although I really don't like calling it Lambda though, because this Task would be used for Batch as well.  So not sure what to call the base class, but like the idea of having one and then adding a Feeder class (which also does not need to be a Lambda).

**Comment by jkeifer** [comment]
In my mind this was only for use for tasks within a cirrus project, i.e., for things packaged as a lambda. But you're suggesting it could be used for tasks built into docker containers as well? I suppose that makes sense.

Then maybe we want to have a `LambdaHandler` mixin that does the handler registration, and we could expose both a `Task` and `TaskLambda` class? Or perhaps a `Lambda` class decorator that does the registration?

I don't know what is best here, I just want to make sure it is super easy for a task author to ensure the task class will actually work with the default lambda definition.yml config.

**Comment by matthewhanson** [comment]
Oh I see, I was wondering what the value was with the registration.   But yeah, you could use the Task class when creating a new Batch task as well, it provides the CLI for local testing and all the wrapping/fetching/parsing/payload return.

We could also change the default generated definition.yml file for new tasks to not use lambda_function.lambda_handler, which as you know I've never cared for.

**Comment by jkeifer** [comment]
I suppose cirrus could import the .py file in the component dir to inspect it to see if it has a subclass of the base (whatever we're calling it), an explicit `lambda_handler` definition, or any other patterns we would consider supporting here. This seems a little more "magic" than I'd normally like, because it comes from a build-time inspection and isn't something you can "see" externally, but advantage over the current just-set-a-default approach is that we could error the build command if no handler is set in the definition.

I guess we also need to know what .py file to import; that could be handled in a similar discovery-with-error process as above.

Though we also have to consider with the lambda-as-batch runner, to make sure whatever is set here is compatible with the patterns it supports. If we are actually modifying the module state on class instantiation to set a well-known `lambda_handler` value, then we get that compatibility for free.

Also, maybe setting `lambda_handler` in the module namespace isn't a problem for batch-only applications. So perhaps we don't need different behavior for lambda vs not.

File: src/cirrus/lib/task.py [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+        parser = subparsers.add_parser('cirrus', parents=[pparser], help=h, formatter_class=dhf)
+        parser.add_argument('url', help='url (s3 or local) to Cirrus Process Payload')
+
+        # turn Namespace into dictionary
+        pargs = vars(parser0.parse_args(args))
+        # only keep keys that are not None
+        pargs = {k: v for k, v in pargs.items() if v is not None}
+
+        if pargs.get('command', None) is None:
+            parser.print_help()
```
Comment (jkeifer) [inline src/cirrus/lib/task.py:171]: Should this be `pparser`?

File: src/cirrus/lib/task.py [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+
+    def __init__(self: "Task", payload: Dict, local: Optional[bool]=None, workdir: Optional[str]=None):
+        # parse event
+        payload = ProcessPayload.from_event(payload)
+
+        # TODO - use PySTAC here
+        self.original_items = payload.pop('features')
+        self.items = deepcopy(self.original_items)
+
+        self.process_definition = payload.pop('process')
```
Comment (jkeifer) [inline src/cirrus/lib/task.py:34]: For compatibility with workflow chaining, I think you want to use `payload.process` here, as that will figure out the current process definition in the payload for you. Otherwise you could get a list of process definitions.

Comment (jkeifer) [reply src/cirrus/lib/task.py:34]: Well, I say that, but it seems this attribute is used two different ways. In light of that realization, I think maybe you want something like this:

```suggestion
        self.process_definition = payload.process
        self._process_definition = payload.pop('process')
```

Then in `output_payload` you would use `self._process_definition.

That said, I'd be interested in learning more about the need here and what problem(s) this whole pull-it-apart-and-rebuild-later strategy solves for us.

File: src/cirrus/lib/task.py [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+            task.logger.error(err, exc_info=True)
+            raise err
+        finally:
+            # remove work directory if not running locally
+            if task._tmpworkdir:
+                task.logger.debug('Removing work directory %s' % task._workdir)
+                rmtree(task._workdir)
+        
+    @classmethod
+    def parse_args(cls, args):
```
Comment (jkeifer) [inline src/cirrus/lib/task.py:142]: For batch specifically, it could be advantageous to allow users to add subcommands/args in the case where they want to have a single container implement several different commands. Thus one image could be used by multiple job definitions. I'm not sure how this would plug in with the `local` vs `cirrus` idea (it seems like mode of operation is separate from which command is being run).

Maybe this is as simple as using `parser.parse_known_args` instead of `parser.parse_args`, and passing any remainder to `process` for the end user to parser. That said, I've never used it but understand it has some edge cases that can be tricky, so I'm not sure if that would work well or not.

File: src/cirrus/lib/task.py [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+
+    def download_assets(self, assets: Optional[List[str]]=None):
+        """Download provided asset keys for all items in payload. Assets are saved in workdir in a
+           directory named by the Item ID
+
+        Args:
+            assets (Optional[List[str]], optional): List of asset keys to download. Defaults to all assets.
+        """
+        for i, item in enumerate(self.items):
+            outdir = op.join(self._workdir, item['id'])
```
Comment (jkeifer) [inline src/cirrus/lib/task.py:75]: It seems like this should probably be like `op.join(self._workdir, item['collection'], item['id'])`.

As an aside, I highly recommend switching from `os.path` to using `pathlib.Path` objects.

File: src/cirrus/lib/task.py (hunk 1 above)
Comment (matthewhanson) [reply src/cirrus/lib/task.py:171]: Actually should be parser0, the root parser.  But I moved out the parser build into it's own function separate from parsing the args, so users can now get the parser and add additional subcommands or args to it

File: src/cirrus/lib/task.py (hunk 2 above)
Comment (matthewhanson) [reply src/cirrus/lib/task.py:34]: Yeah, good points here.

The motivation here was to try and keep things clear for developers as to what should be mutable and what shouldn't. 

The input is items (features) which we want to store the original version, but then also a separate output version of that users would modify (they would never modify the original items, it is for reference and adding source links and such).   I'm also wondering if it makes sense to copy the input items by default into the output structure.

The process definition on the other hand should generally not be changed by the user (although I've hit a couple use cases where I modify the parameters for future tasks).

So basically the reason for splitting them up is because the user interacts with the Items and the Process definition in different ways. Although it probably makes sense to treat them in a similar way.  Make a deepcopy of the original paylod then a copy that can be changed by the user.

File: src/cirrus/lib/task.py (hunk 3 above)
Comment (matthewhanson) [reply src/cirrus/lib/task.py:142]: Refactored the code so that creating the parser is it's own function now that users can fetch then add on it.  They would then redefine the `cli` function.

File: src/cirrus/lib/task.py (hunk 4 above)
Comment (matthewhanson) [reply src/cirrus/lib/task.py:75]: Will switch to pathlib

**Comment by matthewhanson** [comment]
@jkeifer The testing error is also happening in your PR #46 , which is in the transfer tests.  Any idea what could be happening here?

**Comment by matthewhanson** [comment]
I added some tests, although there are still failing tests from test_transfer unrelated to these additions

File: src/cirrus/lib/task.py [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+    def output_options(self) -> Dict:
+        return self.process_definition.get('output_options', {})
+
+    def validate(self) -> bool:
+        # put validation logic on input Items and process definition here
+        return True
+
+    def download_assets(self, assets: Optional[List[str]]=None):
+        """Download provided asset keys for all items in payload. Assets are saved in workdir in a
+           directory named by the Item ID
```
Comment (gadomski) [inline src/cirrus/lib/task.py:107]:
```suggestion
        """Download provided asset keys for all items in payload. Assets are saved in workdir in a
           directory named by the Item ID, and the items are updated with the new asset hrefs.
```

File: src/cirrus/lib/task.py [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+        # The original items from the payload
+        self.original_items = deepcopy(payload['features'])
+
+        self.items = self._payload['features']
+
+        # set up logger
+        self.logger = get_task_logger(f"task.{self._name}", payload=payload)
+
+        if not skip_validation:
+            self.validate()
```
[the remark this answers is not in the document]
Comment (matthewhanson) [reply src/cirrus/lib/task.py:50]: makes total sense, validate can be a classmethod that takes in the whole input payload.

**Comment by matthewhanson** [comment]
I was trying to fix the tests, started changing from mocking to vcr when I updated all the dev dependencies and the tests all work now.   So ready for review so we can get this merged into main @jkeifer

File: CHANGELOG.md [hunk 7]
```diff
@@ -6,6 +6,10 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 
 ## [Unreleased]
 
+### Added
+
+* Abstract base class `Task` for creating new tasks ([#42])
```
Comment (jkeifer) [inline CHANGELOG.md:11]: You need to add the link reference at the bottom.

File: src/cirrus/lib/task.py [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+
+    def upload_assets(self, assets: Optional[List[str]]=None):
+        if self._local:
+            self.logger.warn('Running in local mode, assets not uploaded')
+            return
+        for i, item in enumerate(self.items):
+            self.items[i] = upload_item_assets(item, **self.output_options)
+
+    # this should be in PySTAC
+    @classmethod
```
Comment (jkeifer) [inline src/cirrus/lib/task.py:126]: `@staticmethod`? Looks like `self` isn't used.

File: src/cirrus/lib/task.py [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+                self._name: self._version
+            }
+        self._payload['features'] = self.items
+        # add provenance metadata
+
+        return self._payload
+
+    @property
+    def process_definition(self) -> Dict:
+        return self._payload['process']
```
[the remark this answers is not in the document]
Comment (jkeifer) [reply src/cirrus/lib/task.py:91]: It might be worth adding a workflow chaining test case.

File: src/cirrus/lib/task.py [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+
+# types
+PathLike = Union[str, Path]
+
+
+class Task(ABC):
+
+    _name = 'task'
+    _description = 'A task for doing things'
+    _version = '0.1.0'
```
[the remark this answers is not in the document]
Comment (matthewhanson) [reply src/cirrus/lib/task.py:30]: They are static attributes and this was the easiest way to define them, but yes they should be required. Every base class should define them, so open to options on how to better do that.

File: src/cirrus/lib/task.py (hunk 8 above)
Comment (matthewhanson) [reply src/cirrus/lib/task.py:123]: Good call - changed it to staticmethod, but as noted I think this is best implemented in PySTAC as a general way to create new derived items from existing ones.

File: src/cirrus/lib/task.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (matthewhanson) [reply src/cirrus/lib/task.py]: Updated.
Test for workflow chaining test case sounds like a good idea.  Not it.

**Comment by matthewhanson** [comment]
This work has been replaced by new library [stac-task](https://github.com/stac-utils/stac-task)
